gremlin-driver/src/main/java/org/apache/tinkerpop/gremlin/driver/Client.java

```diff
@@ -530,16 +531,23 @@ protected void initializeImplementation() {
                         .toArray(CompletableFuture[]::new))
                         .join();
             } catch (CompletionException ex) {
-                Throwable cause;
-                Throwable result = ex;
-                if (null != (cause = ex.getCause())) {
-                    result = cause;
-                }
-
-                logger.error("", result);
+                logger.error("", (ex.getCause() == null) ? ex : ex.getCause());
             } finally {
                 hostExecutor.shutdown();
             }
+
+            // throw an error if there is no host available after initializing connection pool.
+            if (cluster.availableHosts().isEmpty()) {
+                throw new NoHostAvailableException();
+            }
+
+            // other-wise we will continue on available hosts and try to re-initiate any unavailable hosts in the background
+            final List<Host> unavailableHosts = cluster.allHosts()
+                    .stream().filter(host -> !host.isAvailable()).collect(Collectors.toList());
+            if (!unavailableHosts.isEmpty()) {
+                CompletableFuture.runAsync(() -> handleUnavailableHosts(unavailableHosts));
```

Comment:
I think you should pass in the existing executor here into this method instead of creating one.

Reply:
Okay, so pass it here and close it in the finally here?

Reply:
Yeah. RunAsync can take in an executor I think.

Creating the Executor and Thread are pretty resource intensive IIRC.